libs/core/langchain_core/messages/utils.py

```diff
@@ -242,3 +248,604 @@ def convert_to_messages(
+    can't be merged.
+
+    Args:
+        messages: Sequence Message-like objects to merge.
+
+    Returns:
+        List of BaseMessages with consecutive runs of message types merged into single
+        messages. If two messages being merged both have string contents, the merged
+        content is a concatenation of the two strings with a new-line separator. If at
+        least one of the messages has a list of content blocks, the merged content is a
+        list of content blocks.
+
+    Example:
+        .. code-block:: python
+
+            from langchain_core.messages import (
+                merge_message_runs,
+                AIMessage,
+                HumanMessage,
+                SystemMessage,
+                ToolCall,
+            )
+
+            messages = [
+                SystemMessage("you're a good assistant."),
+                HumanMessage("what's your favorite color", id="foo",),
+                HumanMessage("wait your favorite food", id="bar",),
+                AIMessage(
+                    "my favorite colo",
+                    tool_calls=[ToolCall(name="blah_tool", args={"x": 2}, id="123")],
+                    id="baz",
+                ),
+                AIMessage(
+                    [{"type": "text", "text": "my favorite dish is lasagna"}],
+                    tool_calls=[ToolCall(name="blah_tool", args={"x": -10}, id="456")],
+                    id="blur",
+                ),
+            ]
+
+            merge_message_runs(messages)
+
+        .. code-block:: python
+
+            [
+                SystemMessage("you're a good assistant."),
+                HumanMessage("what's your favorite color\nwait your favorite food", id="foo",),
+                AIMessage(
+                    [
+                        "my favorite colo",
+                        {"type": "text", "text": "my favorite dish is lasagna"}
+                    ],
+                    tool_calls=[
+                        ToolCall({"name": "blah_tool", "args": {"x": 2}, "id": "123"),
+                        ToolCall({"name": "blah_tool", "args": {"x": -10}, "id": "456")
+                    ]
+                    id="baz"
+                ),
+            ]
+
+    """  # noqa: E501
+    if not messages:
+        return []
+    messages = convert_to_messages(messages)
+    merged: List[BaseMessage] = []
+    for msg in messages:
+        curr = msg.copy(deep=True)
+        last = merged.pop() if merged else None
+        if not last:
+            merged.append(curr)
+        elif isinstance(curr, ToolMessage) or not isinstance(curr, last.__class__):
+            merged.extend([last, curr])
+        else:
+            last_chunk = _msg_to_chunk(last)
+            curr_chunk = _msg_to_chunk(curr)
+            if isinstance(last_chunk.content, str) and isinstance(
+                curr_chunk.content, str
+            ):
+                last_chunk.content += "\n"
```

Comment:
note: we add a newline separator when merging two string messages, which is a more opinionated choice